legion/engine/physics/physics_contact.hpp

```diff
@@ -121,73 +154,178 @@ namespace legion::physics
 
-			math::vec3 Ra, Rb, minRaCrossN, RbCrossN;
-			calculateJacobianComponents(Ra, minRaCrossN, Rb, RbCrossN);
+            //--------------------------- pre calculate contact constraint effective mass -----------------------------------//
 
-			auto rbRef = rbRefHandle.read();
-			auto incRef = rbIncHandle.read();
+            tangentNormal1 = math::cross(collisionNormal, math::vec3(1, 0, 0));
 
-			float maUnit = rbRefHandle ? rbRef.inverseMass : 0.0f;
-			float mbUnit = rbIncHandle ? incRef.inverseMass : 0.0f;
+            if (math::epsilonEqual(math::length(tangentNormal1),0.0f,0.01f))
+            {
+                tangentNormal1 = math::cross(collisionNormal, math::vec3(0, 1, 0));
+            }
 
-			//calculate M-1
-			math::mat3 ma = math::mat3(maUnit);
-			math::mat3 Ia = rbRefHandle ? rbRef.globalInverseInertiaTensor : math::mat3(0.0f);
-			math::mat3 mb = math::mat3(mbUnit);
-			math::mat3 Ib = rbIncHandle ? incRef.globalInverseInertiaTensor : math::mat3(0.0f);
+            tangentNormal1 = math::normalize(tangentNormal1);
 
-			//calculate M^-1 J^T
-			math::vec3 jx = ma * -collisionNormal;
-			math::vec3 jy = Ia * minRaCrossN;
-			math::vec3 jz = mb * collisionNormal;
-			math::vec3 jw = Ib * RbCrossN;
+            tangentNormal2 = math::normalize(math::cross(collisionNormal, tangentNormal1));
 
-			//calculate effective mass
-			float efMx = math::dot(-collisionNormal, jx);
-			float efMy = math::dot(minRaCrossN, jy);
-			float efMz = math::dot(collisionNormal, jz);
-			float efMw = math::dot(RbCrossN, jw);
+            effectiveMass = calculateEffectiveMassOnNormal(collisionNormal);
+            tangent1EffectiveMass = calculateEffectiveMassOnNormal(tangentNormal1);
+            tangent2EffectiveMass = calculateEffectiveMassOnNormal(tangentNormal2);
+		}
 
-			effectiveMass = efMx + efMy + efMz + efMw;
+        float calculateJVConstraint(const math::vec3& normal)
+        {
+            math::vec3 Ra, Rb, minRaCrossN, RbCrossN;
+            calculateJacobianComponents(normal, Ra, minRaCrossN, Rb, RbCrossN);
 
-		}
+            auto RefRB = rbRefHandle.read();
+            auto IncRB = rbIncHandle.read();
+
+            math::vec3 va, wa, vb, wb;
+
+            va = RefRB.velocity;
+            wa = RefRB.angularVelocity;
+            vb = IncRB.velocity;
+            wb = IncRB.angularVelocity;
+
+            float JVx = math::dot(-normal, va);
+            float JVy = math::dot(minRaCrossN, wa);
+            float JVz = math::dot(normal, vb);
+            float JVw = math::dot(RbCrossN, wb);
+
+            return -(JVx + JVy + JVz + JVw);
+        }
+
+        float calculateEffectiveMassOnNormal(math::vec3& normal)
+        {
+            math::vec3 Ra, Rb, minRaCrossN, RbCrossN;
+            calculateJacobianComponents(normal, Ra, minRaCrossN, Rb, RbCrossN);
+
+            auto rbRef = rbRefHandle.read();
+            auto incRef = rbIncHandle.read();
+
+            float maUnit = rbRefHandle ? rbRef.inverseMass : 0.0f;
```

Comment:
Shouldn't the rbRefHandle be tested before reading?

Reply:
iirc, reading a null component handle gives the default values of the component. But that does mean that checking 
rbRefHandle? is actually not necessary. I'll fix it in my next pull request. Thanks ^^